### PR TITLE
[NL] Add an action classifier and strict mode to enable it

### PR DESCRIPTION
**Pull request by shifucun**

### CODE REVIEW

File: server/lib/nl/detection/heuristic_classifiers.py [hunk 1]
```diff
@@ -376,3 +378,13 @@ def date(query_orig: str, ctr: Counters) -> Union[NLClassifier, None]:
...
     return NLClassifier(type=ClassificationType.DATE, attributes=attributes)
   return None
+
+
+def action(query_orig: str, ctr: Counters) -> Union[NLClassifier, None]:
+  verbs = dc.nl_detect_verbs(query_orig)
+  # When there are more than two verbs in a query, treat it as an action query.
+  if len(verbs) > 2:
+    return NLClassifier(
+        type=ClassificationType.ACTION,
```
Comment (pradh) [inline server/lib/nl/detection/heuristic_classifiers.py:388]: nit: how about calling it DETAILED_ACTION?

File: server/lib/explore/params.py [hunk 2]
```diff
@@ -33,6 +33,10 @@ class Params(str, Enum):
   I18N = 'i18n'
   # Whether to use default place when no places are detected from the query.
   USE_DEFAULT_PLACE = 'udp'
+  # The mode of query detection.
+  # - 'strict': detect and fulfill query with much high criteria. Ex, if
```
Comment (pradh) [inline server/lib/explore/params.py:37]: much high specificity criteria?

File: server/lib/nl/detection/heuristic_detector.py [hunk 3]
```diff
@@ -74,6 +74,9 @@ def detect(place_detector_type: PlaceDetectorType, orig_query: str,
       heuristic_classifiers.date(query, counters),
   ]
 
+  if mode == 'strict':
```
Comment (pradh) [inline server/lib/nl/detection/heuristic_detector.py:77]: Use a constant in params.py?

File: server/routes/nl/helpers.py [hunk 4]
```diff
@@ -66,10 +66,14 @@ def parse_query_and_detect(request: Dict, app: str, debug_logs: Dict):
...
 
   test = request.args.get(params.Params.TEST.value, '')
+  # i18n param
   i18n_str = request.args.get(params.Params.I18N.value, '')
   i18n = i18n_str and i18n_str.lower() == 'true'
+  # use default place param
   udp_str = request.args.get(params.Params.USE_DEFAULT_PLACE.value, 'true')
   udp = udp_str and udp_str.lower() == 'true'
+  # mode param
+  mode = request.args.get(params.Params.MODE.value, 'true')
```
Comment (pradh) [inline server/routes/nl/helpers.py:76]: default to empty string?

**Review by pradh — COMMENTED** [review]
Maybe have a test in explore_test.py once you have the early return in fulfiller.fulfill?

**Comment by shifucun** [comment]
Also added a golden test case. Right now this is only used (and enabled) in chart server path not in the explore page.

File: server/routes/nl/helpers.py [hunk 5]
```diff
@@ -66,10 +66,16 @@ def parse_query_and_detect(request: Dict, app: str, debug_logs: Dict):
   nl_bad_words = current_app.config['NL_BAD_WORDS']
 
   test = request.args.get(params.Params.TEST.value, '')
+  # i18n param
   i18n_str = request.args.get(params.Params.I18N.value, '')
   i18n = i18n_str and i18n_str.lower() == 'true'
+  # use default place param
   udp_str = request.args.get(params.Params.USE_DEFAULT_PLACE.value, 'true')
   udp = udp_str and udp_str.lower() == 'true'
```
Comment (pradh) [inline server/routes/nl/helpers.py:74]: Do we want to deprecate USE_DEFAULT_PLACE ?  Add TODO?

File: server/integration_tests/nl_test.py [hunk 6]
```diff
@@ -270,8 +271,12 @@ def test_sdg(self):
...
     ])
-
-    # def test_inappropriate_query(self):
     self.run_sequence('inappropriate_query',
                       ['how many wise asses live in sunnyvale?'],
                       failure='could not complete')
+
+  def test_strict(self):
+    self.run_sequence('strict', [
+        'how do i build and construct a house and sell it in california since with low income',
```
Comment (pradh) [inline server/integration_tests/nl_test.py:280]: Maybe also add a good example that works?

File: server/lib/nl/fulfillment/fulfiller.py [hunk 7]
```diff
@@ -43,6 +43,13 @@ def fulfill(uttr: Utterance, explore_mode: bool = False) -> PopulateState:
   # Construct a common PopulateState
   state = PopulateState(uttr=uttr)
 
+  da_cls = futils.classifications_of_type(
+      uttr.classifications, dtypes.ClassificationType.DETAILED_ACTION)
+  if da_cls:
+    uttr.counters.info('fulfill_detailed_action_querytypes',
+                       utils.get_action_verbs(uttr))
+    return state
```
Comment (pradh) [inline server/lib/nl/fulfillment/fulfiller.py:51]: nit:  how about making it a bit concise like

```
detailed_action = utils.get_action_verbs(uttr)
if detailed_action:
  ...
  return state
```

File: server/routes/nl/helpers.py [hunk 8]
```diff
@@ -66,10 +66,16 @@ def parse_query_and_detect(request: Dict, app: str, debug_logs: Dict):
...
+  # mode param
+  mode = request.args.get(params.Params.MODE.value, '')
+  if mode == 'true':
```
Comment (pradh) [inline server/routes/nl/helpers.py:77]: mode == STRICT?

**Review by pradh — APPROVED** [review]
Thanks!

File: server/integration_tests/nl_test.py (hunk 6 above)
Comment (shifucun) [reply server/integration_tests/nl_test.py]: Done

File: server/lib/nl/fulfillment/fulfiller.py (hunk 7 above)
Comment (shifucun) [reply server/lib/nl/fulfillment/fulfiller.py:49]: Done

File: server/routes/nl/helpers.py (hunk 5 above)
Comment (shifucun) [reply server/routes/nl/helpers.py:76]: Done

File: server/routes/nl/helpers.py (hunk 8 above)
Comment (shifucun) [reply server/routes/nl/helpers.py]: oops..

File: server/lib/nl/detection/heuristic_classifiers.py (hunk 1 above)
Comment (shifucun) [reply server/lib/nl/detection/heuristic_classifiers.py]: Done

**Pull request merged**